### PR TITLE
l2geth: update timestamp logic

### PR DESCRIPTION
**Pull request by tynes**
**Description**
This commit updates the timestamp updating logic
such that `time.Now` is used instead of relying on
L1 timestamps. This gives a higher fidelity for the
`TIMESTAMP` opcode as well as makes the time on L2
be closer to the time on L1.

L1 to L2 transactions no longer have the property of
having the same timestamp on L2 as the timestamp
of the L1 block they were included in.

This should be able to be turned on without needing
hardfork logic as replicas should always accept the
timestamp that the sequencer sets. The sequencer is
a trusted entity in the existing implementation and
it is expected that the sequencer will become more
trustless in future iterations of the protocol.

This change is added to improve both UX and devex.
Users are confused by the timestamps on Etherscan
being ~15 minutes behind. This is due to the timestamps
being set based on L1 block numbers, and the system
only pulls L1 data once a secure amount of PoW
is placed on top. Developers would like the timestamps
to have a higher fidelity and be closer to the timestamps
on L1.

**Metadata**
- Fixes INT-379
- Fixes INT-445

### CODE REVIEW

**Comment by tynes** [comment]
This is now ready for review and includes tests. I need to double check two things before this is ready to merge:
- This code can run as a replica on mainnet now without a network split
- This code can run as a sequencer on a devnet with an older node (pre this PR) can run as a replica without a network split

File: l2geth/rollup/sync_service_test.go [hunk 1]
```diff
@@ -25,71 +25,154 @@ import (
...
+		err = service.applyTransactionToTip(tx3)
+	}()
+	event3 := <-txCh
+	if service.GetLatestL1BlockNumber() != latest {
+		t.Fatal("block number went backwards")
 	}
-	if service.OVMContext == *unexpectedCtx {
-		t.Fatal("context should not be updated because not enough time passed")
+
+	// TODO: should this be malleated?
```
Comment (tynes) [inline l2geth/rollup/sync_service_test.go:173]: TODO: spec out the defined behavior for the L1 blocknumber as it is now decoupled from the timestamp

**Comment by tynes** [comment]
- Try to trigger a timestamp inconsistency by spamming a network

File: l2geth/rollup/sync_service_test.go (hunk 1 above)
Comment (tynes) [reply l2geth/rollup/sync_service_test.go:173]: This behavior should be unchanged so no it should not be malleated

File: integration-tests/test/ovmcontext.spec.ts [hunk 2]
```diff
@@ -74,9 +74,11 @@ describe('OVM Context: Layer 2 EVM Context', () => {
       const l1BlockNumber = await OVMContextStorage.l1BlockNumbers(i)
       expect(l1BlockNumber.toNumber()).to.deep.equal(l1Block.number)
 
-      // L1 and L2 blocks will have the same timestamp.
+      // L1 and L2 blocks will have approximately the same timestamp.
```
Comment (smartcontracts) [inline integration-tests/test/ovmcontext.spec.ts:77]: Can we make the assertion of "strictly greater than" in addition to "approximately equal to"?

File: l2geth/rollup/sync_service.go [hunk 3]
```diff
@@ -435,10 +435,10 @@ func (s *SyncService) SequencerLoop() {
 			log.Error("Could not sequence", "error", err)
 		}
 		s.txLock.Unlock()
+	}
 
-		if err := s.updateContext(); err != nil {
-			log.Error("Could not update execution context", "error", err)
-		}
+	if err := s.updateContext(); err != nil {
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:440]: Why is this moved outside of the loop?

File: l2geth/rollup/sync_service.go [hunk 4]
```diff
@@ -435,10 +435,10 @@ func (s *SyncService) SequencerLoop() {
...
+		log.Error("Could not update execution context", "error", err)
 	}
 }
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:444]: Just leaving myself a mental note that we'll need to update the documentation to reflect this new behavior.

File: l2geth/rollup/sync_service.go [hunk 5]
```diff
@@ -798,31 +797,53 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
...
+	shouldMalleateTimestamp := !s.verifier && tx.QueueOrigin() == types.QueueOriginL1ToL2
+	if tx.L1Timestamp() == 0 || shouldMalleateTimestamp {
+		// Get the latest known timestamp
+		current := time.Unix(int64(ts), 0)
+		// Get the current clocktime
+		now := time.Now()
+		// If enough time has passed, then assign the
+		// transaction to have the timestamp now. Otherwise,
+		// use the current timestamp
+		if now.Sub(current) > s.timestampRefreshThreshold {
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:824]: What is this threshold set to? How often can we expect timestamps to refresh?

Comment (smartcontracts) [reply l2geth/rollup/sync_service.go:824]: Why even have a threshold at all?

Comment (smartcontracts) [reply l2geth/rollup/sync_service.go:824]: If we are required to have a threshold, it's worth noting that we're using the same threshold for both this logic and for the L1 block number updating logic. If we set the threshold too low we may have monotonicity issues with the L1 block number (wouldn't be good).

File: l2geth/rollup/sync_service.go [hunk 6]
```diff
@@ -798,31 +797,53 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
...
+			current = now
+		}
+		tx.SetL1Timestamp(uint64(current.Unix()))
+	} else if tx.L1Timestamp() == 0 && s.verifier {
+		// This should never happen
+		log.Error("No tx timestamp found when running as verifier", "hash", tx.Hash().Hex())
 	} else if tx.L1Timestamp() < s.GetLatestL1Timestamp() {
+		// This should never happen, but sometimes does
 		log.Error("Timestamp monotonicity violation", "hash", tx.Hash().Hex())
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:833]: Unrelated but since I'm seeing it, do we keep track of these errors @optimisticben?

File: l2geth/rollup/sync_service.go [hunk 7]
```diff
@@ -798,31 +797,53 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
...
 	}
 
+	l1BlockNumber := tx.L1BlockNumber()
+	// Set the L1 blocknumber
+	if l1BlockNumber == nil {
+		tx.SetL1BlockNumber(bn)
+	} else if l1BlockNumber.Uint64() > s.GetLatestL1BlockNumber() {
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:840]: Presumably if this `else` clause fails we have a monotonicity violation, should that be logged too?

File: l2geth/rollup/sync_service.go [hunk 8]
```diff
@@ -798,31 +797,53 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
...
+		s.SetLatestL1BlockNumber(l1BlockNumber.Uint64())
+	}
+
+	// Store the latest timestamp value
+	s.SetLatestL1Timestamp(tx.L1Timestamp())
```
Comment (smartcontracts) [inline l2geth/rollup/sync_service.go:845]: Are we sure we want to set this in all cases? Because we still get to this point even if `tx.L1Timestamp() == 0 && s.verifier` or `tx.L1Timestamp() < s.GetLatestL1Timestamp()` gets triggered. Maybe we only want to set if `tx.L1Timestamp() > ts`.

File: l2geth/rollup/sync_service.go (hunk 3 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:440]: Good catch, that is totally a typo

File: l2geth/rollup/sync_service.go (hunk 5 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:824]: There is a threshold here because otherwise we would blow up our batch context overhead, basically every tx would have its own context. The idea would be to set the threshold to 10-15 seconds and then all transactions that happen within that range would get the same timestamp

File: l2geth/rollup/sync_service.go (hunk 8 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:845]: So `tx.L1Timestamp() == 0 && s.verifier` should never happen. The only time when `tx.L1Timestamp() == 0` is when a transaction is fresh from a user coming in via RPC

File: l2geth/rollup/sync_service.go [hunk 9]
```diff
@@ -798,31 +797,53 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
...
-		log.Debug("Updating OVM context based on new transaction", "timestamp", ts, "blocknumber", tx.L1BlockNumber().Uint64(), "queue-origin", tx.QueueOrigin())
+
+	// The L1Timestamp is 0 for QueueOriginSequencer transactions when
+	// running as the sequencer, the transactions are coming in via RPC.
+	// This code path also runs for replicas/verifiers so any logic involving
+	// `time.Now` can only run for the sequencer. All other nodes must listen
+	// to what the sequencer says is the timestamp, otherwise there will be a
+	// network split.
+	shouldMalleateTimestamp := !s.verifier && tx.QueueOrigin() == types.QueueOriginL1ToL2
+	if tx.L1Timestamp() == 0 || shouldMalleateTimestamp {
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:816]: An alternative architecture to this design would be to have a goroutine that updates the global value every 10 seconds to `time.Now` and then don't do any updating here and instead just assigning here

File: l2geth/rollup/sync_service.go (hunk 6 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:833]: We do not track these errors, but that is a good idea

File: l2geth/rollup/sync_service.go (hunk 8 above)
Comment (tynes) [reply l2geth/rollup/sync_service.go:851]: I've noted the above via a comment

File: integration-tests/test/ovmcontext.spec.ts (hunk 2 above)
Comment (tynes) [reply integration-tests/test/ovmcontext.spec.ts:77]: Which do you mean should be greater than? - an L1 block can have at highest a timestamp of `time.Now() + 15s` which could be larger than the `time.Now` that the sequencer assigns to a transaction. This only applies if we do not trail L1 by any amount - then the configurable amount that we trail L1 by impacts which is larger

Comment (tynes) [reply integration-tests/test/ovmcontext.spec.ts:77]: I added another test that asserts that the L1 blocknumber and timestamps are ascending, it runs after the stress tests run

File: l2geth/rollup/sync_service.go [hunk 10]
```diff
@@ -599,17 +599,15 @@ func (s *SyncService) GasPriceOracleOwnerAddress() *common.Address {
 
 /// Update the execution context's timestamp and blocknumber
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:600]: The latest L1 blocknumber logic is now updated, it simply will see if the remote tip is greater than the known latest L1 block number and update it if so

File: l2geth/rollup/sync_service.go (hunk 8 above)
Comment (smartcontracts) [reply l2geth/rollup/sync_service.go:852]: Still wondering if we should gate on `tx.L1Timestamp() > s.GetLatestL1Timestamp()`.

**Review by smartcontracts — APPROVED** [review]
LGTM

**Comment by 0xged** [comment]
Any ETA for this release on mainnet ? 😇  🙏

**Comment by Madalosso** [comment]
> Any ETA for this release on mainnet ? innocent pray

ETA for kovan as well?

**Comment by vinta** [comment]
@0xged @Madalosso It's already live on Optimism Kovan. As far as I know, it will be live on mainnet on 25th if everything goes well on testnet.

**Comment by Madalosso** [comment]
My bad @vinta I thought this would make the timestamps unique until I saw the comment (https://github.com/ethereum-optimism/optimism/pull/1982#discussion_r780572153).
As I saw multiple orders with the same timestamp I thought it wasn't live on kovan yet. My bad.
